pkg/datapath/tables/node_address.go

```diff
@@ -0,0 +1,365 @@
+		}
+	}
+
+	// Remove addresses that were not part of the new set.
+	for addr := range existing {
+		if !new.Has(addr) {
+			updated = true
+			n.NodeAddresses.Delete(txn, addr)
+		}
+	}
+
+	if updated {
+		addrs := showAddresses(new)
+		n.Log.WithField("node-addresses", addrs).Info("Node addresses updated")
+		if reporter != nil {
+			reporter.OK(addrs)
+		}
+	}
+}
+
+func (n *nodeAddressController) getAddressesFromDevice(dev *Device) (addrs sets.Set[NodeAddress]) {
+	addrs = sets.New[NodeAddress]()
+
+	if dev.Flags&net.FlagUp == 0 {
+		return
+	}
+
+	// Skip obviously uninteresting devices.
+	// We include the HostDevice as its IP addresses are consider node addresses
+	// and added to e.g. ipcache as HOST_IDs.
+	if dev.Name != defaults.HostDevice {
+		skip := false
+		for _, prefix := range defaults.ExcludedDevicePrefixes {
+			if strings.HasPrefix(dev.Name, prefix) {
+				skip = true
+				break
+			}
+		}
+		if skip {
+			return
+		}
+	}
+
+	// ipv4 and ipv6 are set to true when the primary address is picked.
+	// Used to implement 'NodePort' and 'Primary' flags.
+	ipv4, ipv6 := false, false
+
+	for _, addr := range sortAddresses(dev.Addrs) {
+		// We keep the scope-based address filtering as was introduced
+		// in 080857bdedca67d58ec39f8f96c5f38b22f6dc0b.
+		if addr.Scope > uint8(n.AddressScopeMax) {
+			// Addresses are sorted by scope, so can stop early.
+			break
+		}
+		if addr.Addr.IsLoopback() {
+			continue
+		}
+
+		// Figure out if the address is usable for NodePort.
+		nodePort := false
+		primary := false
+		if dev.Selected && len(n.Config.NodePortAddresses) == 0 {
+			// The user has not specified IP ranges to filter on IPs on which to serve NodePort.
+			// Thus the default behavior is to use the primary IPv4 and IPv6 addresses of each
+			// device.
+			if addr.Addr.Is4() && !ipv4 {
+				ipv4 = true
+				nodePort = true
+				primary = true
+			}
+			if addr.Addr.Is6() && !ipv6 {
+				ipv6 = true
+				nodePort = true
+				primary = true
+			}
+		} else if ip.NetsContainsAny(n.Config.getNets(), []*net.IPNet{ip.IPToPrefix(addr.AsIP())}) {
+			// User specified --nodeport-addresses and this address was within the range.
+			nodePort = true
+			if addr.Addr.Is4() && !ipv4 {
+				primary = true
+				ipv4 = true
+			} else if addr.Addr.Is6() && !ipv6 {
+				primary = true
+				ipv6 = true
+			}
+		}
+
+		addrs.Insert(NodeAddress{Addr: addr.Addr, Primary: primary, NodePort: nodePort, DeviceName: dev.Name})
+	}
+	return
+}
+
+// showAddresses formats a Set[NodeAddress] as "1.2.3.4 (eth0), fe80::1 (eth1)"
+func showAddresses(addrs sets.Set[NodeAddress]) string {
+	ss := make([]string, 0, len(addrs))
+	for addr := range addrs {
+		ss = append(ss, addr.String())
+	}
+	sort.Strings(ss)
+	return strings.Join(ss, ", ")
+}
+
+// sortAddresses returns a copy of the addresses, sorted by primary (e.g. !iIFA_F_SECONDARY) and then by
+// address scope.
+func sortAddresses(addrs []DeviceAddress) []DeviceAddress {
```

Comment:
nit: mildly surprising to see a departure from the typical go sorting interface. It's called out in the comment though, so I'm okay with it, but maybe you can think of a way to work `copy` into the name of the function too?

Reply:
Hmm, this also made me realize one of the big pitfalls of using slices in structs that are stored in statedb. Once there's func iters we might want to add `ImmutableList[T]` or something similar.